### ISSUE TITLE
Bug in assembler resulting in bad code generation

### ISSUE THREAD

**Issue by psyker156** [CommunicationsSecurityEstablishment/spartacus#18]
Hey!
There is a bug in the assembler when using a label as part of an instruction. If the programmer makes a mistake and use ":" after the label, the assembler does not pick up the error and will not generate the proper code resulting in the following error being raised:
`
raise ValueError("Invalid instruction detected at address {}".format(hex(address)))
`

Here is an example of bad code that will cause the error (check the jmp):
```
    mov #1024 $a
    mov translationTable $b
createTable:
    memw [4] #0x0 $b
    add #4 $b
    sub #1 $a
    cmp #0 $a
    jmp <l> createTable:
breakHere:
    memw [4] #0x8001000 $b
```

**Comment by kommisar5150** [comment]
Technically, you could name your label with a colon at the end (eg. createtable::) and it would still work, since the assembler's parser looks at the last character to determine if there's a label mnemonic. 

For example, the following code is valid:

`.global start

start:
    MOV end $S
    MOV text $A
    MOV #11 $B
    MOV #0x20001000 $D
loop:
    MEMR [1] $A $C
    MEMW [1] $C $D
    ADD #1 $A
    ADD #1 $D
    SUB #1 $B
    CMP #0 $B
    JMP <LH> loop
endlessloop::
    JMP <> endlessloop:
text:
    .dataAlpha test :loop: 
end:`

Though it would be an interesting question whether or not we should allow this to be accepted. Thoughts?

**Comment by cannotparse** [comment]
Considering this is an assembly language, I don't think `:` is a fitting character for the label names. Don't know what other assemblers allow, so if others use it then sure.

---

So I tried running @kommisar5150 's code and noticed it behaved oddly. 

It was printing `:` in the front of the string. I tried making the string bigger by adding a few aaa's to see what was wrong. Turns out it was being placed at the first address of the string. 
[Here's a gist of me debugging it](https://gist.github.com/cannotparse/ff7a3c490af24801d8e7c8e8648c0ef5). 

I wanted to make his code better reflect @psyker156 's code snippet with the lower case 'l' in the conditions to be thorough and I got a different result (I actually stumbled on this first). 

I was getting 'ä'(ASCII code `0x84`, showed as a '__' when printed and the debugger won't print it) prepended to the printed string. I double checked that the printing was done properly (Thanks for the code snippet @kommisar5150 ). I decided to debug it afterwards to see what was going on. [Here's a gist with my memory dump + changes](https://gist.github.com/cannotparse/75ba8dfc98b6c6b6eb560c168a14b3d6). 

Let me know if it's behaving like this for you. Could be my outdated version of spartacus but I doubt that.

---
This part I'm looking for the bug and hunting for it. The gists just have all the data when through and I found. Mostly here for the curious.

My hunch was the extra `:` was causing issues somewhere in the assembler's parser and was getting placed somewhere else. So I went looking for the `0x84` to see if I accidentally inserted it somewhere in the file without me noticing. The hexdumps show this character doesn't exist in the file (see gist) 🙄.

So while I was looking into the binary of these files, I wanted to figure out what the compiled difference was. I know @pg-cse mentioned an issue with the xml generated by the linker back at cusec.

Only difference I found was ([gist here](https://gist.github.com/cannotparse/9bfa48e31aaf3f385ca60febc47ed85b))
```
<<<< my version without :labels
^@::`:ERR_MSG>
----- 
^@:`:ERR_MSG>
>>>> my version with :labels
``` 
and 
```
<<<< my version without :labels
^@:test>
----
^@::test>
>>>> my version with :labels
```

Seems like it will confuse either the code generated by the assembler for its symbols or the loader can't read it properly. Either way, its a bug with how the parser handles the bad state.

----

Since I think @pg-cse would be able to fix the issue a lot faster than me; I'm posting my findings before I get into the code. To summarize for everyone else until this is fixed:
- @psyker156 's code is working with a new feature he's trying to add and has a lot more complicated assembly to work with it. Based off what I've seen; labels and instructions for his code were affected by the bad state in the paser (unless he introduced a bug with his feature.)
- @kommisar5150 's code is short and sweet and doesn't fully demonstrate the bug. 

TL:DR - Don't use `:` in labels, it might break, it might not. Do you want the risk it?

**Comment by cannotparse** [comment]
I've made a hotfix(aka a bandaid) to prevent adding additional `:`. Still looking for the core reason this is causing undefined behaviour. I'll be making a PR soonish. Also, [x86 doesn't allow it](https://docs.oracle.com/cd/E26502_01/html/E28388/eqbsx.html#esqaq) see this for [identifiers](https://docs.oracle.com/cd/E26502_01/html/E28388/eqbsx.html#eoqjw).

**Comment by kommisar5150** [comment]
I tested @cannotparse 's labeltest.casm file and I too had a` : `extra added in front of the text, which was 100% caused by the label having an added colon. This happens in both Windows and Linux so it doesn't seem to be a compatibility issue. The assembler replaces the label's colons when it finds one

`buildInstruction.instructionCode = sline[0].replace(MEMORY_REFERENCE_INDICATORS, "").upper()`

So I'm not sure how it ends up in the dataAlpha field. Curiously, if I place two print loops to print the same text twice in a row, the second print omits the colon at the beginning but still places what seems to be either a space or null character. I'll see if I can trace it.

**Comment by cannotparse** [comment]
@kommisar5150 The hotfix I was going to propose was to following:

```python3
        # ISSUE #18 HOTFIX - Avoid multiple ':' in a label
        if COMMENT_INDICATORS == line[0][0]:
            # Kept in hotfix to maintain operation until better solution found
            buildInstruction.instructionCode = line[0].replace(MEMORY_REFERENCE_INDICATORS, "").upper()
            return

        if (MEMORY_REFERENCE_INDICATORS == line[0][-1]):
            # This is no code line... No need for further work, return!
            # We keep the comment indicator so we can later discard the instruction.
            # Put text into instructionCode since this is an impossible case... Easy to
            # follow in calling code. And put in some indicator for the linker.
            # ISSUE #18 HOTFIX - Avoid multiple ':' in a label
            if line[0].count(':') > 1:
                raise ValueError("Syntax error, label can't contain multiple {} ".format(MEMORY_REFERENCE_INDICATORS))

            buildInstruction.instructionCode = line[0].replace(MEMORY_REFERENCE_INDICATORS, "").upper()
            return

```

**Comment by kommisar5150** [comment]
That would definitely work to enforce naming conventions for labels, I'm all for it. Though it still irks me that it seems to be labels used as immediate values that's messing up the dataAlpha strings.

```
.global start

start:
    MOV end $S
    MOV text $A
    MOV #15 $B
    MOV #0x20001000 $D
loop:
    MEMR [1] $A $C
    MEMW [1] $C $D
    ADD #1 $A
    ADD #1 $D
    SUB #1 $B
    CMP #0 $B
    JMP <LH> loop
endlessloop::
    JMP <> endlessloop
text:
    .dataAlpha test :loop:aaaa
end:
```

In this example, if we remove the colon from `JMP <> endlessloop`, it prints no problem. Might need a fix to prevent immediate labels to contain colons as well.

**Comment by cannotparse** [comment]
@kommisar5150 So based on the snippet of code we both pointed out, it should (in theory), strip out the extra colon from the label. Which is why my symbol tables don't include the colon. So the but must be with how other opts handle memrefs.

**Comment by cannotparse** [comment]
Just went step through the assembler. Essentially, the assembler is working as intended. There isn't a bad state introduced. The problem lies with how the linker/debugger load the program into memory. You can read the [spec](https://github.com/CommunicationsSecurityEstablishment/spartacus/blob/master/Documentation/SpartacusReference-EN.md) about how the assembled program is supposed to look like. It must be caused by the extra `:`. I'm still looking into what it actually does to the other tools and how it affects the binary. [Here's some more data from pdb xD](https://gist.github.com/cannotparse/d359601030fffafb2b8bbc2744c00ccd)

**Comment by cannotparse** [comment]
I figured out the root cause of why we are getting odd behaviour. I was right in the summary in my first comment, about the undefined behaviour. Have a look at [this gist](https://gist.github.com/cannotparse/525714660927d6ce099b8a340d573a06#file-jiberish-stuff-i-joted-down) for where I got the conclusion from.  

Essentially, when the instruction is parsed by the parser it also creates its future location/offset based on the expected length. However, in the lines mentioned below, the possibility that these aren't in fact that length gets ignored. In the examples we created to test, this leads to something seemingly harmless since the next sections are alphaData. In the situation @psyker156 was in, it was probably other instructions. That lead to the capua core throwing the exception when it doesn't recognize what's happening.

That was fun.

lines as mentioned 
- [Assembler/Parser/Parser.py:443](https://github.com/CommunicationsSecurityEstablishment/spartacus/blob/master/ToolChain/Assembler/Parser/Parser.py#L443)
- [Assembler/Parser/Parser.py:448](https://github.com/CommunicationsSecurityEstablishment/spartacus/blob/master/ToolChain/Assembler/Parser/Parser.py#L448)
- [Assembler/Parser/Parser.py:137](https://github.com/CommunicationsSecurityEstablishment/spartacus/blob/master/ToolChain/Assembler/Parser/Parser.py#L137)
- [Assembler/Parser/Parser.py:470](https://github.com/CommunicationsSecurityEstablishment/spartacus/blob/master/ToolChain/Assembler/Parser/Parser.py#L470)

Basically, 443 & 448 wrap the additional `:`, 470 uses the length from a hardcoded value from [FormDescription.py:32](https://github.com/CommunicationsSecurityEstablishment/spartacus/blob/master/CapuaEnvironment/IntructionFetchUnit/FormDescription.py#L31) and stores that into the instruction object. 137 trusts it and updates the offsets. 

There are plenty if the way you can tackle this but I'd recommend just removing `:` from the possible labels and have someone redo the parsing/toolchain as good first commits.

**Comment by pg-cse** [comment]
I don't think using ":" in label names is a good thing. As stated by @cannotparse the ":" character is a bit of a "keyword" in most assembly languages.
The way I see it now is, either a hot fix is published for this or not. Either way, I also strongly believe that the full tool chain needs a rewrite: this was written for debugging purpose in like 2 sittings on new year's eve 3 years ago. It was never really meant to be used by programmers but only to make testing of the core easier.

**Comment by pg-cse** [comment]
This was fixed by https://github.com/CommunicationsSecurityEstablishment/spartacus/pull/23
The assembler will now output an error message if the programmer did an error like the one described in this issue.
